feat(api): Add save/load mechanism to serialize/deserialize module calibration data.

vegano1 opened the pull request:
# Overview

Once the module is calibrated we need to be able to save/load this calibration data so it's re-used on startup and we don't have to calibrate all the time. This pull request adds this mechanism and builds off [RCORE-609](https://opentrons.atlassian.net/browse/RCORE-609)

Closes [RCORE-611](https://opentrons.atlassian.net/browse/RCORE-611)

Module calibration data will be stored in `/data/robot/modules/` with the name `{module_id}.json` where the `module_id` is the serial number of the module, for example

```
root@4e01b6:/data/robot/modules# cat TDV21P20210413B05.json

{"offset": [1.0, 2.0, 3.0], "mount": 1, "slot": 3, "module": "temperatureModuleType", "module_id": "TDV21P20210413B05", "instrument_id": "P50SV3320220729A07", "lastModified": "2023-03-21T14:44:28.062314+00:00", "source": "user", "status": {"markedBad": false, "source": null, "markedAt": null}}
```

# Test Plan

- [x] Make sure we can save module calibration data in `/data/robot/module/{module_id}.json`
- [x] Make sure we can load module calibration data stored in `/data/robot/module/`
- [x] Make sure we can delete module calibration data matching `{module_id}.json` from the disk.
- [x] Make sure we module calibration data is stored when `module_calibration` command is called from OT3repl

# Changelog

- Added loading/saving functions that serialize/deserialize a new ModuleOffsetModel

# Review requests

- Is there anything I'm missing?

[RCORE-609]: https://opentrons.atlassian.net/browse/RCORE-609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RCORE-611]: https://opentrons.atlassian.net/browse/RCORE-611?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
feat(api): add module calibration mechanism.

vegano1 committed:
fix comment
DO NOT reset previous instrument offset when calibrating module

vegano1 committed:
prep calibrate_module mechanism so we arent relying on geometric calculations in the hardware controller as that belongs in the protocol engine

vegano1 committed:
fix linting

vegano1 committed:
feat(api): add load/save module calibration offset mechanism.

vegano1 committed:
Merge branch 'RCORE-609_autocalibration_on_arbitrary_position' into RCORE-611_module_calibration_storage

vegano1 committed:
Merge branch 'edge' into RCORE-609_autocalibration_on_arbitrary_position

vegano1 committed:
Merge branch 'RCORE-609_autocalibration_on_arbitrary_position' into RCORE-611_module_calibration_storage

vegano1 committed:
just use serial number as the module_id when loading/saving data

vegano1 committed:
clean up from code review

vegano1 committed:
Merge branch 'edge' into RCORE-609_autocalibration_on_arbitrary_position

vegano1 committed:
comment fixup

vegano1 committed:
Merge branch 'RCORE-609_autocalibration_on_arbitrary_position' into RCORE-611_module_calibration_storage

vegano1 committed:
fix hardware-testing script

vegano1 committed:
fix test script

vegano1 committed:
Merge branch 'RCORE-609_autocalibration_on_arbitrary_position' into RCORE-611_module_calibration_storage

Laura-Danielle commented on api/src/opentrons/calibration_storage/ot3/models/v1.py:
```diff
@@ -82,3 +84,25 @@ class InstrumentOffsetModel(BaseModel):
+class ModuleOffsetModel(BaseModel):
+    offset: Point = Field(..., description="Module offset found from calibration.")
+    mount: OT3Mount = Field(..., description="The mount used to calibrate this module.")
+    slot: int = Field(..., description="The slot this module was calibrated in.")
```
Maybe we should also record the instrument id used to calibrate the module as well as the mount?

Laura-Danielle reviewed (approved):
One comment about saving the pipette id used to calibrate the module offsets as well, otherwise it looks good to me.

vegano1 committed:
rename structure not found error so its more descriptive
we are removing LINEAR_SEARCH  method in a future pr, so change the function signatures so we default to BINARY_SEARCH so its easier to merge.

vegano1 committed:
Merge branch 'RCORE-609_autocalibration_on_arbitrary_position' into RCORE-611_module_calibration_storage

vegano1 committed:
store instrument_id as well

sfoster1 commented on api/src/opentrons/hardware_control/modules/module_calibration.py:
```diff
@@ -0,0 +1,81 @@
+        source=SourceType.default,
+        status=CalibrationStatus(),
+    )
+    if module_id and ff.enable_ot3_hardware_controller():
```
Do we really need this feature flag check? On an OT-2, there just won't be module calibration data to load, and that path is already handled by defaulting to 0.

sfoster1 commented on api/src/opentrons/hardware_control/modules/module_calibration.py:
```diff
@@ -0,0 +1,81 @@
+    instrument_id: Optional[str] = None,
+) -> None:
+    """Save the calibration offset for a given module."""
+    if module_id and ff.enable_ot3_hardware_controller():
```
Do we really need this feature flag check? On an OT-2, we just won't be running this flow.

sfoster1 reviewed (approved):
Looks good, but I'd love to remove those feature flag checks - I think there's not actually any bad behavior that they prevent.

vegano1 committed:
remove feature flag from module_calibration since its a flex only flow

vegano1 committed:
Merge branch 'edge' into RCORE-611_module_calibration_storage

vegano1 committed:
mypy
